bpo-41093: TCPServer's serve_forever() shuts down immediately when calling shutdown()

tontinton opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41093](https://bugs.python.org/issue41093) -->
https://bugs.python.org/issue41093
<!-- /issue-number -->

tontinton committed:
bpo-41093: Add `server_shutdown()` overridable function to BaseServer

server_shutdown() is called on BaseServer's shutdown method to make
server_forever() return immediately on shutdown not waiting for select()
to timeout or return on EVENT_READ.

tontinton committed:
bpo-41093: Implement `server_shutdown()` for TCPServer

Fixed issue `41093` as now every server inheriting TCPServer will
shutdown immediately after calling shutdown().

Note: UDPServer doesn't call shutdown() on it's server socket as there
is no need.

dorosch commented on Lib/socketserver.py:
```diff
@@ -474,6 +477,14 @@ def server_activate(self):
+            self.socket.shutdown(socket.SHUT_RDWR)
+        except OSError:
+            pass  # Don't call shutdown on a bad file descriptor
+        self.server_close()
```
This method mark as overridden but if we override it we need to know that need to call `self.server_close()`. This is not obvious in my opinion.

tontinton replied:
Fixed, from now server_shutdown will only shutdown the socket without closing it.

dorosch commented on Lib/socketserver.py:
```diff
@@ -221,10 +222,6 @@ def serve_forever(self, poll_interval=0.5):
         """
         self.__is_shut_down.clear()
         try:
-            # XXX: Consider using another file descriptor or connecting to the
```
In my opinion, don't need to remove this comment because the proposal has not yet been resolved.

tontinton replied:
I added the comment again, but I do think that by implementing the server_shutdown() function this is resolved

dorosch commented on Lib/socketserver.py:
```diff
@@ -249,6 +246,7 @@ def shutdown(self):
         deadlock.
         """
         self.__shutdown_request = True
+        self.server_shutdown() # selector.select() will return immediately
```
Please follow the comment style of this file. Comments are written before the comment line.

tontinton replied:
Fixed

dorosch commented on Lib/socketserver.py:
```diff
@@ -531,6 +542,10 @@ def server_activate(self):
         # No need to call listen() for UDP.
         pass
 
+    def server_shutdown(self):
```
In my opinion, it is worth leaving this method as `pass`. How this is done with the `server_activate` method since this is an unobvious behavior of the method.

tontinton replied:
Fixed

dorosch reviewed (commented):
Thanks so much for your commits and welcome to the cpython.
Your changes look good. Perhaps you would describe these changes in the documentation before other participants look at your code there. You can added docs for your changes in `Doc/library/sockerserver.rst`. And in my opinion you can added some tests for your changes in `Lib/test/test_sockerserver.py`.

tontinton committed:
bpo-41093: Add docs to BaseServer's server_shutdown()

tontinton committed:
bpo-41093: Add a test for the issue #41093

The test creates a TCPServer and calls server_forever on it with a high
poll interval of 30 seconds, then calls shutdown on it and validates that it took less
than a second.
